File: Src/Engine/Editor/Modules/SceneModule.cpp

#include "SceneModule.h"

#include "SceneGraphModule.h"
#include "Core/Platform/Win32/Win32File.h"
#include "Core/Types/Collections/ListExtensions.h"
#include "Core/Types/Strings/StringView.h"
#include "Runtime/Level/Level.h"
#include "Runtime/Level/Scene/Scene.h"
#include "Runtime/UI/GUI//Rootcontrol.h"

#include "Editor/EditorApp.h"
#include "Editor/GUI/Tree/ActorTreeNode.h"
#include "Editor/SceneGraph/ActorGraphNode.h"
#include "Editor/SceneGraph/Actor/SceneGraphNode.h"

namespace SE::Editor
{
	SceneModule::SceneModule(EditorApp* editor): EditorModule(editor)
	{
	}

	void SceneModule::OnSceneSaving(Scene* scene, const UID &sceneId)
	{
		OnSaveStart(RootControl::GetGameRoot());
	}

	void SceneModule::OnSceneSaved(Scene* scene, const UID &sceneId)
	{
		OnSaveEnd(RootControl::GetGameRoot());
	}

	void SceneModule::OnSceneSaveError(Scene* scene, const UID &sceneId)
	{
		OnSaveEnd(RootControl::GetGameRoot());
	}

	void SceneModule::OnSceneLoaded(Scene* scene, const UID &sceneId)
	{
		DateTime startTime = DateTime::NowUTC();

		// Build scene tree
		SceneGraphNode* sceneNode = editor->sceneGraphModule->BuildSceneTree(scene);
		ActorTreeNode* treeNode = sceneNode->TreeNode;
		treeNode->SetIsLayoutLocked(true);
		treeNode->Expand(true);

		// Add to the tree
		ActorTreeNode* rootNode = Root->TreeNode;
		rootNode->SetIsLayoutLocked(true);
		sceneNode->ParentNode = Root;
		rootNode->SortChildren();
		rootNode->SetIsLayoutLocked(false);
		rootNode->Parent->PerformLayout();

		DateTime endTime = DateTime::NowUTC();
		int milliseconds = (endTime - startTime).GetTotalMilliseconds();
		LOG_INFO("Scene", "Created graph for scene \'{0}\' in {1} ms", scene->GetName(), milliseconds);
	}

	void SceneModule::OnSceneUnloading(Scene* scene, const UID &sceneId)
	{
		// Find scene tree node
		ActorGraphNode* node = Root->FindChildActor(scene);
		if (node != nullptr)
		{
			LOG_INFO("Scene", "Cleanup graph for scene \'{0}\'", scene->GetName());

			// Cleanup
			auto selection = editor->sceneModule->Selection;
			bool hasSceneSelection = false;
			for (int i = 0; i < selection.Count(); i++)
			{
				if (selection[i]->ParentScene == node)
				{
					hasSceneSelection = true;
					break;
				}
			}
			if (hasSceneSelection)
			{
				List<ScenesGraphNode*> newSelection;
				for (int i = 0; i < selection.Count(); i++)
				{
					if (selection[i]->ParentScene != node)
						newSelection.Add(selection[i]);
				}
				editor->sceneModule->Select(newSelection);
			}
			node->Dispose();
		}
	}

	void SceneModule::OnActorSpawned(Actor* actor)
	{
		// Skip for not loaded scenes (spawning actors during scene loading in script Start function)
		ActorGraphNode* sceneNode = GetActorNode(actor->GetScene());
		if (sceneNode == nullptr)
		{
			return;
		}

		// Skip for missing parent
		Actor* parent = actor->GetParent();
		if (parent == nullptr)
		{
			return;
		}

		ActorGraphNode* parentNode = GetActorNode(parent);
		if (parentNode == nullptr)
		{
			// Missing parent node when adding child actor to not spawned or unlinked actor
			return;
		}

		ActorGraphNode* node = editor->sceneGraphModule->BuildActorNode(actor);
		if (node != nullptr)
		{
			node->ParentNode = parentNode;
		}
	}

	void SceneModule::OnActorDeleted(Actor* actor)
	{
		ActorGraphNode* node = GetActorNode(actor);
		if (node != nullptr)
		{
			OnActorDeleted(node);
		}
	}

	void SceneModule::OnActorDeleted(ActorGraphNode* node)
	{
		for (int i = 0; i < node->ChildNodes.Count(); i++)
		{
			ActorGraphNode* child;
			if (TypeTryCast(node->ChildNodes[i], child))
			{
				i--;
				OnActorDeleted(child);
			}
		}

		// ActorRemoved?.Invoke(node);

		// Cleanup part of the graph
		node->Dispose();
	}

	void SceneModule::OnActorParentChanged(Actor* actor, Actor* prevParent)
	{
		ActorGraphNode* node = nullptr;
		ActorGraphNode* parentNode = GetActorNode(actor->GetParent());

		// Try use previous parent actor to find actor node
		ActorGraphNode* prevParentNode = GetActorNode(prevParent);
		if (prevParentNode != nullptr)
		{
			// If should be one of the children
			node = prevParentNode->FindChildActor(actor);

			// Search whole tree if node was not found
			if (node == nullptr)
			{
				node = GetActorNode(actor);
			}
		}
		else if (parentNode != nullptr)
		{
			// Create new node for that actor (user may unlink it from the scene before and now link it)
			node = editor->sceneGraphModule->BuildActorNode(actor);
		}
		if (node == nullptr)
			return;

		// Get the new parent node (may be missing)
		node->ParentNode = parentNode;
		if (parentNode == nullptr)
		{
			// Check if actor is selected in editor
			if (Selection.Contains(node))
			{
				Deselect();
			}

			// Remove node (user may unlink actor from the scene but not destroy the actor)
			node->Dispose();
		}
	}

	void SceneModule::OnActorOrderInParentChanged(Actor* actor)
	{
		ActorGraphNode* node = GetActorNode(actor);
		if (node != nullptr)
		{
			node->TreeNode->OnOrderInParentChanged();
		}
	}

	void SceneModule::OnActorNameChanged(Actor* actor)
	{
		ActorGraphNode* node = GetActorNode(actor);
		if (node != nullptr)
		{
			node->TreeNode->UpdateText();
		}
	}

	void SceneModule::OnActorActiveChanged(Actor* actor)
	{
		ActorGraphNode* node = GetActorNode(actor);
		if (node != nullptr)
		{
			ENGINE_UNREACHABLE_CODE()
			// node->TreeNode->OnActiveChanged();
		}
	}

	List<ScenesGraphNode*>& SceneModule::ScenesRootNode::__GetSelection()
	{
		return _editor.sceneModule->Selection;
	}

	void SceneModule::Select(Actor* actor)
	{
		ActorGraphNode* node = GetActorNode(actor);
		if (node != nullptr)
		{
			Select(node);
		}
	}

	void SceneModule::Select(List<ScenesGraphNode*>& selection, bool additive)
	{
		if (selection.IsEmpty())
		{
			Deselect();
			return;
		}

		// Prevent from selecting null nodes
		for (int i = selection.Count() - 1; i >= 0; i--)
		{
			if (selection[i] == nullptr)
			{
				selection.RemoveAt(i);
			}
		}


		// Check if won't change
		if (!additive && Selection.Count() == selection.Count())
		{
			auto equalFunc = CreateFunc([](ScenesGraphNode* &a, ScenesGraphNode* &b)
			{
				return a == b;
			});

			if (ListExtensions::SequenceEqual(Selection, selection, equalFunc))
			{
				return;
			}
		}


		if (!additive)
		{
			Selection.Clear();
		}
		Selection.Add(selection);

		SelectionChanged();
	}

	void SceneModule::Select(ScenesGraphNode* selection, bool additive)
	{
		if (selection == nullptr)
		{
			return;
		}

		// Check if won't change
		if (!additive && Selection.Count() == 1 && Selection[0] == selection)
		{
			return;
		}
		if (additive && Selection.Contains(selection))
		{
			return;
		}

		if (!additive)
		{
			Selection.Clear();
		}
		Selection.Add(selection);

		// SelectionChange(Selection);
	}

	void SceneModule::Duplicate()
	{
		/*// Peek things that can be copied (copy all actors)
		List<ScenesGraphNode*> nodes = Selection.Where(x => x.CanDuplicate).ToList().BuildAllNodes();
		if (nodes.Count == 0)
			return;
		List<Actor*> actors = List<Actor*>();
		List<SceneGraphNode*> newSelection = List<SceneGraphNode*>();
		// List<IUndoAction> customUndoActions = null;
		for (ScenesGraphNode* node : nodes)
		{
			if (node->CanDuplicate)
			{
				ActorGraphNode* actorNode;
				if (TypeTryCast(node, actorNode))
				{
					actors.Add(actorNode->Actor);
				}
				else
				{
					var customDuplicatedObject = node->Duplicate(out var customUndoAction);
					if (customDuplicatedObject != null)
						newSelection.Add(customDuplicatedObject);
					if (customUndoAction != null)
					{
						if (customUndoActions == null)
							customUndoActions = new List<IUndoAction>();
						customUndoActions.Add(customUndoAction);
					}
				}
			}
		}

		if (actors.Count() == 0)
		{
			// Duplicate custom scene graph nodes only without actors
			if (newSelection.Count != 0)
			{
				// Select spawned objects (parents only)
				var selectAction = new SelectionChangeAction(Selection.ToArray(), newSelection.ToArray(), OnSelectionUndo);
				selectAction.Do();

				// Build a single compound undo action that pastes the actors, pastes custom stuff (scene graph extension) and selects the created objects (parents only)
				var customUndoActionsCount = customUndoActions?.Count ?? 0;
				var undoActions = new IUndoAction[1 + customUndoActionsCount];
				for (int i = 0; i < customUndoActionsCount; i++)
					undoActions[i] = customUndoActions[i];
				undoActions[undoActions.Length - 1] = selectAction;

				Undo.AddAction(new MultiUndoAction(undoActions));
				OnSelectionChanged();
			}
			return;
		}

		// Serialize actors
		var data = Actor.ToBytes(actors.ToArray());
		if (data == null)
		{
			Editor.LogError("Failed to copy actors data.");
			return;
		}

		// Create paste action (with selecting spawned objects)
		var pasteAction = PasteActorsAction.Duplicate(data, Guid.Empty);
		if (pasteAction != null)
		{
			pasteAction.Do(out _, out var nodeParents);

			// Select spawned objects (parents only)
			newSelection.AddRange(nodeParents);
			var selectAction = new SelectionChangeAction(Selection.ToArray(), newSelection.ToArray(), OnSelectionUndo);
			selectAction.Do();

			// Build a single compound undo action that pastes the actors, pastes custom stuff (scene graph extension) and selects the created objects (parents only)
			var customUndoActionsCount = customUndoActions?.Count ?? 0;
			var undoActions = new IUndoAction[2 + customUndoActionsCount];
			undoActions[0] = pasteAction;
			for (int i = 0; i < customUndoActionsCount; i++)
				undoActions[i + 1] = customUndoActions[i];
			undoActions[undoActions.Length - 1] = selectAction;

			Undo.AddAction(new MultiUndoAction(undoActions));
			OnSelectionChanged();
		}

		// Scroll to new selected node while duplicating
		Editor.Windows.SceneWin.ScrollToSelectedNode();*/
	}

	void SceneModule::Delete()
	{
		// Peek things that can be removed
		/*var objects = Selection.Where(x => x.CanDelete).ToList().BuildAllNodes().Where(x => x.CanDelete).ToList();
		if (objects.Count == 0)
			return;
		var isSceneTreeFocus = Editor.Windows.SceneWin.ContainsFocus;

		SelectionDeleteBegin?.Invoke();

		// Change selection
		var action1 = new SelectionChangeAction(Selection.ToArray(), new SceneGraphNode[0], OnSelectionUndo);

		// Delete objects
		var action2 = new DeleteActorsAction(objects);

		// Merge two actions and perform them
		var action = new MultiUndoAction(new IUndoAction[]
		{
			action1,
			action2
		}, action2.ActionString);
		action.Do();
		Undo.AddAction(action);

		SelectionDeleteEnd?.Invoke();

		if (isSceneTreeFocus)
		{
			Editor.Windows.SceneWin.Focus();
		}

		// fix scene window layout
		Editor.Windows.SceneWin.PerformLayout();
		Editor.Windows.SceneWin.PerformLayout();*/
	}

	void SceneModule::Copy()
	{
		// Peek things that can be copied (copy all actors)
		/*var objects = Selection.Where(x => x.CanCopyPaste).ToList().BuildAllNodes().Where(x => x.CanCopyPaste && x is ActorNode).ToList();
		if (objects.Count == 0)
			return;

		// Serialize actors
		var actors = objects.ConvertAll(x => ((ActorNode)x).Actor);
		var data = Actor.ToBytes(actors.ToArray());
		if (data == null)
		{
			Editor.LogError("Failed to copy actors data.");
			return;
		}

		// Copy data
		Clipboard.RawData = data;*/
	}

	void SceneModule::PasteActor(Actor* pasteTargetActor)
	{
		// Get clipboard data
		/*var data = Clipboard.RawData;

		// Set paste target if only one actor is selected and no target provided
		if (pasteTargetActor == null && SelectionCount == 1 && Selection[0] is ActorNode actorNode)
		{
			pasteTargetActor = actorNode.Actor.Scene == actorNode.Actor ? actorNode.Actor : actorNode.Actor.Parent;
		}

		// Create paste action
		var pasteAction = PasteActorsAction.Paste(data, pasteTargetActor?.ID ?? Guid.Empty);
		if (pasteAction != null)
		{
			pasteAction.Do(out _, out var nodeParents);

			// Select spawned objects (parents only)
			var selectAction = new SelectionChangeAction(Selection.ToArray(), nodeParents.Cast<SceneGraphNode>().ToArray(), OnSelectionUndo);
			selectAction.Do();

			// Build single compound undo action that pastes the actors and selects the created objects (parents only)
			Undo.AddAction(new MultiUndoAction(pasteAction, selectAction));
			OnSelectionChanged();
		}

		// Scroll to new selected node while pasting
		Editor.Windows.SceneWin.ScrollToSelectedNode();*/
	}

	void SceneModule::Spawn(Actor* actor, Actor* parent, bool autoSelect)
	{
		bool isPlayMode = false;// Editor.StateMachine.IsPlayMode;

		if (Level::IsAnySceneLoaded() == false)
		{
			LOG_ERROR("Scene", "Cannot spawn actor when no scene is loaded.");
			return;
		}

		// SpawnBegin?.Invoke();

		// Add it
		Level::SpawnActor(actor, parent);

		// Peek spawned node
		ActorGraphNode* actorNode = GetActorNode(actor);
		// "Failed to create scene node for the spawned actor."
		ENGINE_ASSERT(actorNode != nullptr);

		// During play in editor mode spawned actors should be dynamic (user can move them)
		/*if (isPlayMode)
			actor.StaticFlags = StaticFlags.None;*/

		// Call post spawn action (can possibly setup custom default values)
		// actorNode->PostSpawn();

		// Create undo action
		/*IUndoAction action = new DeleteActorsAction(actorNode, true);
		if (autoSelect)
		{
			var before = Selection.ToArray();
			Selection.Clear();
			Selection.Add(actorNode);
			OnSelectionChanged();
			action = new MultiUndoAction(action, new SelectionChangeAction(before, Selection.ToArray(), OnSelectionUndo));
		}
		Undo.AddAction(action);*/

		// Mark scene as dirty
		/*MarkSceneEdited(actor->Scene);

		SpawnEnd?.Invoke();

		OnDirty(actorNode);*/
	}

	ActorGraphNode* SceneModule::GetActorNode(Actor* actor)
	{
		if (actor == nullptr)
			return nullptr;

		return GetActorNode(actor->GetInstanceID());
	}

	ActorGraphNode* SceneModule::GetActorNode(UID actorId)
	{
		// ActorNode has the same ID as actor does
		ScenesGraphNode* sceneGraphNode = editor->sceneGraphModule->FindNode(actorId);
		ActorGraphNode* actorGraphNode = nullptr;
		if (!TypeTryCast<ActorGraphNode>(sceneGraphNode, actorGraphNode))
		{
			actorGraphNode = nullptr;
		}
		return actorGraphNode;
	}

	void SceneModule::Deselect()
	{
		// Check if won't change
		if (Selection.Count() == 0)
			return;

		List<ScenesGraphNode*> before = Selection;
		Selection.Clear();

		// SelectionChange(before);
	}

	void SceneModule::OnInit()
	{
		Root = New<ScenesRootNode>();
		Root->LinkTreeNode();

		// Bind events
		Level::SceneSaving.Bind<SceneModule, &SceneModule::OnSceneSaving>(this);
		Level::SceneSaved.BindUnique<SceneModule, &SceneModule::OnSceneSaved>(this);
		Level::SceneSaveError.BindUnique<SceneModule, &SceneModule::OnSceneSaveError>(this);
		Level::SceneLoaded.BindUnique<SceneModule, &SceneModule::OnSceneLoaded>(this);
		Level::SceneUnloading.BindUnique<SceneModule, &SceneModule::OnSceneUnloading>(this);
		Level::ActorSpawned.BindUnique<SceneModule, &SceneModule::OnActorSpawned>(this);
		Level::ActorDeleted.BindUnique<SceneModule, &SceneModule::OnActorDeleted>(this);
		Level::ActorParentChanged.BindUnique<SceneModule, &SceneModule::OnActorParentChanged>(this);
		Level::ActorOrderInParentChanged.BindUnique<SceneModule, &SceneModule::OnActorOrderInParentChanged>(this);
		Level::ActorNameChanged.BindUnique<SceneModule, &SceneModule::OnActorNameChanged>(this);
		Level::ActorActiveChanged .BindUnique<SceneModule, &SceneModule::OnActorActiveChanged>(this);
	}

	void SceneModule::OnExit()
	{
		// Unbind events
		Level::SceneSaving.Unbind<SceneModule, &SceneModule::OnSceneSaving>(this);
		Level::SceneSaved.Unbind<SceneModule, &SceneModule::OnSceneSaved>(this);
		Level::SceneSaveError.Unbind<SceneModule, &SceneModule::OnSceneSaveError>(this);
		Level::SceneLoaded.Unbind<SceneModule, &SceneModule::OnSceneLoaded>(this);
		Level::SceneUnloading.Unbind<SceneModule, &SceneModule::OnSceneUnloading>(this);
		Level::ActorSpawned.Unbind<SceneModule, &SceneModule::OnActorSpawned>(this);
		Level::ActorDeleted.Unbind<SceneModule, &SceneModule::OnActorDeleted>(this);
		Level::ActorParentChanged.Unbind<SceneModule, &SceneModule::OnActorParentChanged>(this);
		Level::ActorOrderInParentChanged.Unbind<SceneModule, &SceneModule::OnActorOrderInParentChanged>(this);
		Level::ActorNameChanged.Unbind<SceneModule, &SceneModule::OnActorNameChanged>(this);
		Level::ActorActiveChanged .Unbind<SceneModule, &SceneModule::OnActorActiveChanged>(this);
	}

	void SceneModule::CreateSceneFile(StringView path)
	{
		Scene* scene = New<Scene>();

		/*var cam = scene->AddChild<Camera>();
		cam.Name = "Camera";
		cam.Position = new Vector3(0, 150, -300);*/

		List<byte> bytes;
		if (!Level::SaveSceneToBytes(bytes, scene))
		{
			LOG_ERROR("Level", "Serialize Scene File failed");
		}

		if (!File::WriteAllBytes(path, bytes))
		{
			LOG_ERROR("Level", "Create Scene File failed");
		}
	}

	void SceneModule::SaveScene(Scene* scene)
	{
		ActorGraphNode* node = GetActorNode(scene);
		SceneGraphNode* sceneNode = static_cast<SceneGraphNode*>(node);
		SaveScene(sceneNode);
	}

	void SceneModule::SaveScene(SceneGraphNode* scene)
	{
		if (!scene->IsEdited)
			return;

		scene->IsEdited = false;
		Level::SaveSceneAsync(scene->scene);
	}

	void SceneModule::OpenScene(UID sceneId, bool additive)
	{
		/*if (!Editor.StateMachine.CurrentState.CanChangeScene)
			return;*/

		// In play-mode Editor mocks the level streaming script
		/*if (Editor.IsPlayMode)
		{
			if (!additive)
			{
				Level.UnloadAllScenesAsync();
			}

			Level::LoadSceneAsync(sceneId);
			return;
		}*/

		/*if (!additive)
		{
			// Ensure to save all pending changes
			if (CheckSaveBeforeClose())
				return;
		}*/

		// Load scene
		Level::LoadSceneAsync(sceneId);
	}

	void SceneModule::OnSaveStart(ContainerControl* uiRoot)
	{
		// Force viewport UI to have fixed size during scene/prefabs saving to result in stable data (less mess in version control diffs)
		if (m_UiRootSizes == nullptr)
		{
			m_UiRootSizes = New<Dictionary<ContainerControl*, Float2>>();
		}
		m_UiRootSizes->At(uiRoot) = uiRoot->Size;
		uiRoot->Size = Float2(1920, 1080);
	}

	void SceneModule::OnSaveEnd(ContainerControl* uiRoot)
	{
		// Restore cached size of the UI root container
		Float2 size;
		if (m_UiRootSizes != nullptr && m_UiRootSizes->TryGet(uiRoot, size))
		{
			uiRoot->Size = size;
			m_UiRootSizes->Remove(uiRoot);
		}
	}

	int SceneModule::InitOrder()
	{
		return -91;
	}

	int SceneModule::__GetSelectionCount()
	{
		return Selection.Count();
	}

	bool SceneModule::__GetHasSthSelected()
	{
		return Selection.Count() > 0;
	}
} // SE

File: Src/Engine/Editor/Modules/SceneModule.cs
using System;
using System.Collections.Generic;
using SE.Editor.SceneGraph;

namespace SE.Editor
{
    /// <summary>
    /// Owns loaded-scene editor state, scene-graph synchronization, and selection.
    /// </summary>
    public sealed class SceneModule : EditorModule
    {
        /// <summary>
        /// Root node for all loaded scenes and their actor hierarchies.
        /// </summary>
        public sealed class ScenesRootNode : RootGraphNode
        {
            private readonly SceneModule m_Module;

            internal ScenesRootNode(SceneModule module)
                : base(module.SceneGraph)
            {
                m_Module = module;
            }

            public override IReadOnlyList<ScenesGraphNode> Selection => m_Module.SelectedNodes;

            public override void Spawn(SE.Actor actor, SE.Actor? parent)
            {
                // The native root hook is a Flax placeholder whose forwarding
                // call is commented out. Keep this override intentionally inert.
                _ = actor;
                _ = parent;
            }
        }

        private readonly List<SceneEntry> m_Scenes = new();
        private readonly List<int> m_SceneSelection = new();
        private readonly List<ScenesGraphNode> m_NodeSelection = new();

        public SceneModule(Editor editor)
            : base(editor)
        {
            SceneGraph = new SceneGraphFactory();
            Root = new ScenesRootNode(this);
            SceneGraph.SetRoot(Root);
            Root.LinkTreeNode();
        }

        public event Action? ScenesChanged;
        public event Action? SelectionChanged;
        public event Action? HierarchyChanged;

        public SceneGraphFactory SceneGraph { get; }
        public ScenesRootNode Root { get; }
        public IReadOnlyList<SceneEntry> Scenes => m_Scenes;
        public IReadOnlyList<ScenesGraphNode> SelectedNodes => m_NodeSelection;
        public IReadOnlyList<int> Selection => m_SceneSelection;
        public int SelectionCount => m_NodeSelection.Count != 0 ? m_NodeSelection.Count : m_SceneSelection.Count;
        public bool HasSelection => SelectionCount != 0;

        internal override int Order => -91;

        public void Select(int sceneIndex, bool additive = false)
        {
            if (sceneIndex < 0 || sceneIndex >= m_Scenes.Count)
                return;

            bool changed = false;
            if (!additive && m_NodeSelection.Count != 0)
            {
                m_NodeSelection.Clear();
                changed = true;
            }
            if (!additive && (m_SceneSelection.Count != 1 || m_SceneSelection[0] != sceneIndex))
            {
                m_SceneSelection.Clear();
                changed = true;
            }
            if (!m_SceneSelection.Contains(sceneIndex))
            {
                m_SceneSelection.Add(sceneIndex);
                changed = true;
            }

            if (changed)
                SelectionChanged?.Invoke();
        }

        public void Select(SE.Actor actor)
        {
            ActorGraphNode? node = GetActorNode(actor);
            if (node != null)
                Select(node);
        }

        public void Select(ScenesGraphNode node, bool additive = false)
        {
            ArgumentNullException.ThrowIfNull(node);
            if (node.IsDisposed)
                return;
            Select(new[] { node }, additive);
        }

        public void Select(IEnumerable<ScenesGraphNode> nodes, bool additive = false)
        {
            ArgumentNullException.ThrowIfNull(nodes);
            List<ScenesGraphNode> requested = new();
            foreach (ScenesGraphNode node in nodes)
            {
                if (node != null && !node.IsDisposed && !requested.Contains(node))
                    requested.Add(node);
            }

            if (requested.Count == 0)
            {
                Deselect();
                return;
            }

            if (!additive && SequenceEqual(m_NodeSelection, requested) && m_SceneSelection.Count == 0)
                return;

            bool changed = false;
            if (!additive)
            {
                changed = m_SceneSelection.Count != 0 || !SequenceEqual(m_NodeSelection, requested);
                m_SceneSelection.Clear();
                m_NodeSelection.Clear();
            }
            foreach (ScenesGraphNode node in requested)
            {
                if (!m_NodeSelection.Contains(node))
                {
                    m_NodeSelection.Add(node);
                    changed = true;
                }
            }

            if (changed)
                SelectionChanged?.Invoke();
        }

        public void Deselect()
        {
            if (m_SceneSelection.Count == 0 && m_NodeSelection.Count == 0)
                return;

            m_SceneSelection.Clear();
            m_NodeSelection.Clear();
            SelectionChanged?.Invoke();
        }

        public bool SaveScene(int sceneIndex)
        {
            return sceneIndex >= 0 && sceneIndex < m_Scenes.Count && SE.Level.SaveSceneAt(sceneIndex);
        }

        public void SaveScene(SE.Scene scene)
        {
            if (GetActorNode(scene) is SceneGraphNode sceneNode)
                SaveScene(sceneNode);
        }

        public void SaveScene(SceneGraphNode sceneNode)
        {
            ArgumentNullException.ThrowIfNull(sceneNode);
            if (!sceneNode.IsEdited)
                return;

            sceneNode.IsEdited = false;
            SE.Level.SaveSceneAsync(sceneNode.Scene);
        }

        public void Spawn(SE.Actor actor, SE.Actor? parent = null, bool autoSelect = true)
        {
            ArgumentNullException.ThrowIfNull(actor);
            if (!SE.Level.IsAnySceneLoaded)
            {
                SE.Debug.LogError("Cannot spawn actor when no scene is loaded.");
                return;
            }

            bool failed = parent == null ? SE.Level.SpawnActor(actor) : SE.Level.SpawnActor(actor, parent);
            if (failed)
            {
                SE.Debug.LogError($"Failed to spawn actor '{actor.Name}'.");
                return;
            }

            // Native automatic selection and undo are still commented out.
            _ = autoSelect;
        }

        public ActorGraphNode? GetActorNode(SE.Actor? actor)
        {
            return actor == null ? null : GetActorNode(actor.SceneObjectId);
        }

        public ActorGraphNode? GetActorNode(Guid actorId)
        {
            return SceneGraph.FindNode(actorId) as ActorGraphNode;
        }

        public override void OnInit()
        {
            SE.Level.SceneLoaded += OnSceneLoaded;
            SE.Level.SceneUnloading += OnSceneUnloading;
            SE.Level.SceneUnloaded += OnSceneUnloaded;
            SE.Level.ActorSpawned += OnActorSpawned;
            SE.Level.ActorDeleted += OnActorDeleted;
            SE.Level.ActorParentChanged += OnActorParentChanged;
            SE.Level.ActorOrderInParentChanged += OnActorOrderInParentChanged;
            SE.Level.ActorNameChanged += OnActorNameChanged;
            SE.Level.ActorActiveChanged += OnActorActiveChanged;

            BuildLoadedScenes();
            RefreshSceneEntries();
        }

        public override void OnDispose()
        {
            SE.Level.SceneLoaded -= OnSceneLoaded;
            SE.Level.SceneUnloading -= OnSceneUnloading;
            SE.Level.SceneUnloaded -= OnSceneUnloaded;
            SE.Level.ActorSpawned -= OnActorSpawned;
            SE.Level.ActorDeleted -= OnActorDeleted;
            SE.Level.ActorParentChanged -= OnActorParentChanged;
            SE.Level.ActorOrderInParentChanged -= OnActorOrderInParentChanged;
            SE.Level.ActorNameChanged -= OnActorNameChanged;
            SE.Level.ActorActiveChanged -= OnActorActiveChanged;

            SceneGraph.Dispose();
            m_Scenes.Clear();
            m_SceneSelection.Clear();
            m_NodeSelection.Clear();
            ScenesChanged = null;
            SelectionChanged = null;
            HierarchyChanged = null;
        }

        private void BuildLoadedScenes()
        {
            for (int index = 0; index < SE.Level.ScenesCount; index++)
            {
                SE.Scene scene = SE.Level.GetScene(index);
                if (scene != null && SceneGraph.FindNode(scene.SceneObjectId) == null)
                    AddSceneTree(scene);
            }
        }

        private void AddSceneTree(SE.Scene scene)
        {
            SceneGraphNode sceneNode = SceneGraph.BuildSceneTree(scene);
            sceneNode.ParentNode = Root;
            sceneNode.TreeNode.Expand(true);
        }

        private void OnSceneLoaded(SE.Scene scene, Guid sceneId)
        {
            if (scene != null && SceneGraph.FindNode(sceneId) == null)
                AddSceneTree(scene);

            RefreshSceneEntries();
            HierarchyChanged?.Invoke();
        }

        private void OnSceneUnloading(SE.Scene scene, Guid sceneId)
        {
            ScenesGraphNode? node = SceneGraph.FindNode(sceneId);
            if (node == null && scene != null)
                node = SceneGraph.FindNode(scene.SceneObjectId);
            if (node == null)
                return;

            bool selectionChanged = RemoveSelectionForTree(node);
            node.Dispose();
            if (selectionChanged)
                SelectionChanged?.Invoke();
            HierarchyChanged?.Invoke();
        }

        private void OnSceneUnloaded(SE.Scene scene, Guid sceneId)
        {
            RefreshSceneEntries();
        }

        private void OnActorSpawned(SE.Actor actor)
        {
            if (actor == null || GetActorNode(actor.Scene) == null)
                return;

            SE.Actor parentActor = actor.Parent;
            if (parentActor == null)
                return;

            ActorGraphNode? parentNode = GetActorNode(parentActor);
            if (parentNode == null)
                return;

            // Parent-change and spawn notifications can be adjacent. Reuse a
            // node already created by the first notification.
            ActorGraphNode node = GetActorNode(actor) ?? SceneGraph.BuildActorNode(actor);
            node.ParentNode = parentNode;
            HierarchyChanged?.Invoke();
        }

        private void OnActorDeleted(SE.Actor actor)
        {
            ActorGraphNode? node = GetActorNode(actor);
            if (node == null)
                return;

            bool selectionChanged = RemoveSelectionForTree(node);
            node.Dispose();
            if (selectionChanged)
                SelectionChanged?.Invoke();
            HierarchyChanged?.Invoke();
        }

        private void OnActorParentChanged(SE.Actor actor, SE.Actor previousParent)
        {
            ActorGraphNode? node = GetActorNode(actor);
            ActorGraphNode? parentNode = GetActorNode(actor.Parent);
            ActorGraphNode? previousParentNode = GetActorNode(previousParent);

            if (previousParentNode != null)
                node = previousParentNode.FindChildActor(actor) ?? GetActorNode(actor);
            else if (node == null && parentNode != null)
                node = SceneGraph.BuildActorNode(actor);

            if (node == null)
                return;

            node.ParentNode = parentNode;
            if (parentNode == null)
            {
                bool selectionChanged = RemoveSelectionForTree(node);
                node.Dispose();
                if (selectionChanged)
                    SelectionChanged?.Invoke();
            }

            HierarchyChanged?.Invoke();
        }

        private void OnActorOrderInParentChanged(SE.Actor actor)
        {
            ActorGraphNode? node = GetActorNode(actor);
            node?.TreeNode.OnOrderInParentChanged();
            if (node != null)
                HierarchyChanged?.Invoke();
        }

        private void OnActorNameChanged(SE.Actor actor)
        {
            ActorGraphNode? node = GetActorNode(actor);
            node?.RefreshActor(actor);
            if (node != null)
                HierarchyChanged?.Invoke();
        }

        private void OnActorActiveChanged(SE.Actor actor)
        {
            // The native active-state tree update is still a commented Flax
            // placeholder, so the managed migration intentionally does nothing.
            _ = actor;
        }

        private bool RemoveSelectionForTree(ScenesGraphNode root)
        {
            HashSet<ScenesGraphNode> removed = new();
            CollectTree(root, removed);
            int previousCount = m_NodeSelection.Count;
            m_NodeSelection.RemoveAll(removed.Contains);
            return previousCount != m_NodeSelection.Count;
        }

        private static void CollectTree(ScenesGraphNode node, HashSet<ScenesGraphNode> result)
        {
            if (!result.Add(node))
                return;
            foreach (ScenesGraphNode child in node.ChildNodes)
                CollectTree(child, result);
        }

        private void RefreshSceneEntries()
        {
            int count = SE.Level.ScenesCount;
            bool changed = count != m_Scenes.Count;
            List<SceneEntry> refreshed = new(count);
            for (int index = 0; index < count; index++)
            {
                SceneEntry entry = new(index, SE.Level.GetSceneName(index));
                refreshed.Add(entry);
                changed |= index >= m_Scenes.Count || !m_Scenes[index].Equals(entry);
            }

            if (!changed)
                return;

            m_Scenes.Clear();
            m_Scenes.AddRange(refreshed);
            m_SceneSelection.RemoveAll(index => index < 0 || index >= m_Scenes.Count);
            ScenesChanged?.Invoke();
            SelectionChanged?.Invoke();
        }

        private static bool SequenceEqual(IReadOnlyList<ScenesGraphNode> left, IReadOnlyList<ScenesGraphNode> right)
        {
            if (left.Count != right.Count)
                return false;
            for (int index = 0; index < left.Count; index++)
            {
                if (!ReferenceEquals(left[index], right[index]))
                    return false;
            }

            return true;
        }
    }

    public readonly struct SceneEntry : IEquatable<SceneEntry>
    {
        public SceneEntry(int index, string name)
        {
            Index = index;
            Name = name ?? string.Empty;
        }

        public int Index { get; }
        public string Name { get; }

        public bool Equals(SceneEntry other)
        {
            return Index == other.Index && string.Equals(Name, other.Name, StringComparison.Ordinal);
        }

        public override bool Equals(object? obj)
        {
            return obj is SceneEntry other && Equals(other);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Index, Name);
        }
    }
}
